### ISSUE TITLE
Create autosaves whenever an option dialogue is encountered 

### ISSUE THREAD

**Issue by cooljeanius** [wesnoth/wesnoth#5118]
### Game and System Information
 - What version of the game are you running?
1.14.12
 - Where did you download it from?
Steam
 - What OS are you running?
macOS Catalina

### Describe the desired feature
I would like for the game created an autosave whenever an `[option][/option]` block is encountered. Often times those are at ends of turns (i.e., far away from the start-of-turn autosave) and you can't really prepare for them ahead of time unless you've read the code and know they're coming and create a manual save. That way you could have an easy way to select between the other branches presented in the option.

**Comment by gfgtdf** [comment]
Even if we would agree that this is a useful idea, we probably couldn't implement this since [option]s appear in the middle of an event and the game cannot save the game during an event since it cannot describe the state of an event exceution in a savefile.

That said, i don't really think its a good idea, if whether you can win a scenario depends on whether you took the right option in a choice you don't know nothing about it probably not a very good scenario design, but chances are that the choice was not that much important anyway,

Lastly: you can, if you want, save the game after the [option], then open the savefile and check 'show replay', then replay the game until the point you want to continue it (before the option), and then click 'continue play', its a little work but it allows you to to continue a game from any earlier point of a scenario (unless the scenario designer wrote bad wml which causes OOS errors)

**Comment by stevecotton** [comment]
There are related discussions in #3703 and #4164. An interesting point was raised in https://github.com/wesnoth/wesnoth/issues/4164#issuecomment-529073166 - sometimes the player isn't aware that there was a branch.

**Comment by soliton-** [comment]
Maybe a more realistic idea would be to allow the scenario designer to trigger a save where they deem useful. Either explicitely or silently (though not sure how useful the latter would be since the player might never notice).

**Comment by cooljeanius** [comment]
> Maybe a more realistic idea would be to allow the scenario designer to trigger a save where they deem useful. Either explicitely or silently (though not sure how useful the latter would be since the player might never notice).

This would also be a reasonable solution.

**Comment by CelticMinstrel** [comment]
> it cannot describe the state of an event exceution in a savefile.

This would need to be solved for either solution. If the designer can trigger a save at any time, then that could be (and inevitably will be) in the middle of an event, so the game needs to be able to continue the event when you load that save.

I don't think it's conceptually that difficult to represent an event in progress in the save file, but I have no idea if there are difficulties in implementing the event resumption on load.

If the designer can trigger a save at any time, then we could easily add a call to `wesnoth.create_save()` or whatever in the `[message]` implementation, using the second solution to implement the original request. I'm not convinced this is a good idea, mind you – I believe message options are often used for things that don't really count as a branch. But it would be possible.

**Comment by soliton-** [comment]
Well, we don't have to allow saves at any arbitrary point (though if you figure out how to do that that might be nice as well). Just allowing it in WML where it is already permitted in the GUI seems fine. Not sure how easy that is to describe or understand for the average UMC author though. I didn't look up what conditions actually have to be met. Just have to be outside of an event?

**Comment by gfgtdf** [comment]
No its more complicated than that, and actually  unrelated to [event]s more to 'player actions', for a (random) example: you also cannot save the game between the healing of two units during the start of a turn becasue they game wouldn't remember which units are healed (this is store in a local (c++) variable of the healing code not in the gamestate object.)

**Comment by stevecotton** [comment]
I think the UX should let the player make the branching choice first, and only afterwards offer a savestate with the situation before the choice was made. It's a game dialog that interrupts the flow of character dialogue, so I'd fit in it like this:

* Angry elf: Are you going to let these kidnapping bandits live?
* Hero: Well, let's at least capture them alive and give 'em a fair trial.
* Angry elf: We're abandoning you in this forest right now.
* Player: Crap! Half my units have disappeared
* (half-second pause)
* Game engine: This is a diverging campaign path, and you've just chosen the hard route. Do you want to keep an autosave so that you can play both branches from this point?

That UX design also has an impact on the implementation - it would need the savestate to be stored sometime before prompting the user, so it seems reasonable to require the WML author to set it up before any other actions in the event. We could even prototype it by copying the most recent start-of-turn autosave, with the possibility of improving that later if the feature is popular. It seems reasonable to me to make this feature require autosaves, and be disabled if autosaves are turned off.

**Comment by CelticMinstrel** [comment]
Hmm, not sure if I like that version. I'm not really a fan of the "you just branched, want to save?" idea at all, though I certainly wouldn't mind if that method is supported.

Regardless of how the game decides to actually make a save… if the issue is that you simply cannot represent the game state at certain points (for example in the middle of healing), then I think what the game should do is to _always_ capture a temporary snapshot of the game state whenever it's possible to do so, for example whenever the player has control, at the start of each turn, etc. Then there would be a Lua API function that saves the temporary snapshot to disk. There could also be a separate Lua API function that saves when it next becomes possible, if that's desirable.

**Comment by cooljeanius** [comment]
> > Maybe a more realistic idea would be to allow the scenario designer to trigger a save where they deem useful. Either explicitely or silently (though not sure how useful the latter would be since the player might never notice).
> 
> This would also be a reasonable solution.

Example case where the "allow the scenario designer to trigger a save via WML" solution could be useful: cooljeanius/A_Little_Adventure#33 (there the branching takes place based on killing an enemy, rather than an `[option]` dialogue)

**Comment by knyghtmare** [comment]
the current practice is saving just before the event (if foreknowledge), or reloading the last autosave of the previous scenario and then saving from there before progressing so that both branches can be played.

**Comment by Pentarctagon** [comment]
Would it be more practical to instead have a way to mark/rename the most recent start of turn autosave file in a way to prevent it from getting automatically cleaned up?

**Comment by CelticMinstrel** [comment]
That would work but is a very loose interpretation of this request. After all, you might have moved several units before you triggered the branch point. If you reload from start of turn, you'll need to move those units again.

**Comment by Pentarctagon** [comment]
True, but it seems like it would get 90% of the way there for minimal effort vs needing a lot more effort plus probably having a significant risk of creating buggy save files (option 1) or adding stuttering from having to copy the entire gamestate frequently (for games that would have large save files).

**Comment by CelticMinstrel** [comment]
It's true that it would get 90% of the way there, but I think getting 100% of the way there isn't that much additional effort. It's roughly equivalent to adding an auto-save after every action – move, recruit, recall, attack.

**Comment by Pentarctagon** [comment]
Right, I'm concerned about the performance of that for large saves though. The game freezing for a fraction of a second after each action would be pretty bad.

**Comment by cooljeanius** [comment]
I think getting 90% of the way there is fine; Wesnoth development does have the whole "KISS" principle after all

**Comment by CelticMinstrel** [comment]
> The game freezing for a fraction of a second after each action would be pretty bad.

If you're judging this based on the cost of writing to disk, just making the snapshot in memory might not be quite as bad?

**Comment by Pentarctagon** [comment]
It wouldn't be as bad as writing to disk, but that doesn't meant it definitely wouldn't be noticeable, so it's something that would need to be checked if that's the approach taken.
